### PR TITLE
Add parallelization to CA class

### PR DESCRIPTION
**Pull request by emapco**
Add:
- Add parallelization to CA class with the use of OpenMP pragma directives.
- Update makefiles to compile new targets for the parallelized implementations. 
  - Add OpenMP support for both `Linux` and `Mac OSX`
- Add new Makefile for compiling library object files that combine the CA class object file with the utility functions object file.  

Refactor:
- The `get_state_from_neighborhood` methods now call a function to generate neighborhood array. 
  - There should provide a slight performance boost when generating `Moore` neighborhoods.
- Also make some methods `private` since `User II` shouldn't be able to access them.
- Replace anonymous function with a function defined in `CAutils.h`.
  - Causing issues on some compilers.

### CODE REVIEW

**Comment by TsukiTiger** [comment]
This update resolved the issue of some compiler did not like `lambda` methods as well

File: Source/Datatypes/cellularautomata.cpp [hunk 1]
```diff
@@ -419,8 +419,7 @@ int CellularAutomata::set_new_cell_state(int *cell_index, int index_size,
         }
         // find the max_element in the counter based on the pairs' second variable (number of "votes")
         auto max_elem = max_element(state_votes_counter.begin(), state_votes_counter.end(),
-                                    [](const std::pair<int, int> &a, const std::pair<int, int> &b)
-                                    { return a.second < b.second; });
+                                    less_than_votes);
```
Comment (TsukiTiger) [inline Source/Datatypes/cellularautomata.cpp:425]: Cool Update!

**Review by TsukiTiger — COMMENTED** [review]
Nice Work! Thank you!

**Comment by trevoroldham** [comment]
I pulled this version of the code and I get a compiler error:
g++       -Wall -Wextra -O0 -c -g -fopenmp -DENABLE_OMP -I../../Include cellularautomata.cpp -o cellularautomata_omp.o
clang: error: unsupported option '-fopenmp'

**Pull request merged**